javalib/src/main/scala/java/math/BigDecimal.scala

```diff
@@ -0,0 +1,1729 @@
+    new Array[Long](len -1).scanLeft[Long, Array[Long]](1)((z, e) => z * pow)
+
+  /** Return an increment that can be -1,0 or 1, depending on {@code roundingMode}.
+    *
+    * @param parityBit can be 0 or 1, it's only used in the case {@code HALF_EVEN}
+    * @param fraction the mantissa to be analyzed
+    * @param roundingMode the type of rounding
+    * @return the carry propagated after rounding
+    */
+  private def roundingBehavior(parityBit: Int, fraction: Int,
+      roundingMode: RoundingMode): Int = {
+    import RoundingMode._
+
+    val absFraction = Math.abs(fraction)
+    val sigFraction = java.lang.Integer.signum(fraction)
+    // the carry after rounding
+    roundingMode match {
+      case UP          => sigFraction
+      case DOWN        => 0
+      case CEILING     => Math.max(sigFraction, 0)
+      case FLOOR       => Math.min(sigFraction, 0)
+      case HALF_UP     => if (absFraction >= 5) sigFraction else 0
+      case HALF_DOWN   => if (absFraction > 5) sigFraction else 0
+      case HALF_EVEN   => if (absFraction + parityBit > 5) sigFraction else 0
+      case UNNECESSARY =>
+        if (fraction == 0) 0
+        else throw new ArithmeticException("Rounding necessary")
+    }
+  }
+
+  private def safeLongToInt(longValue: Long): Int = {
+    if (longValue < Int.MinValue || longValue > Int.MaxValue)
+      throw new ArithmeticException("Out of int range: " + longValue)
+
+    longValue.toInt
+  }
+
+  /** The value 0 with the most approximated scale of type {@code int}.
+    *
+    * If {@code longScale > Integer.MAX_VALUE} the scale will be
+    * {@code Integer.MAX_VALUE}.
+    *
+    * If {@code longScale < Integer.MIN_VALUE} the scale will be
+    * {@code Integer.MIN_VALUE}.
+    *
+    * Otherwise {@code longScale} is casted to the type {@code int}.
+    *
+    * @param longScale the scale to which the value 0 will be scaled.
+    * @return the value 0 scaled by the closer scale of type {@code int}.
+    * @see #scale
+    */
+  private def zeroScaledBy(longScale: Long): BigDecimal = {
+    if (longScale == longScale.toInt)
+      valueOf(0, longScale.toInt)
+    else if (longScale >= 0)
+      new BigDecimal(0, Int.MaxValue)
+    else
+      new BigDecimal(0, Int.MinValue)
+  }
+
+  protected def bitLength( sValue: Long): Int = {
+    val smallValue = if (sValue < 0) ~sValue else sValue
+    64 - java.lang.Long.numberOfLeadingZeros(smallValue)
+  }
+
+  private def bitLength(sValue: Int): Int = {
+    val smallValue = if (sValue < 0) ~sValue else sValue
+    32 - java.lang.Integer.numberOfLeadingZeros(smallValue)
+  }
+
+  private def charNotEqualTo(c: Char, cs: Char*): Boolean =
+    cs.forall(ch => c != ch)
+
+  private def charEqualTo(c: Char, cs: Char*): Boolean = cs.contains(c)
+}
+
+class BigDecimal() extends Number with Comparable[BigDecimal] {
+  import BigDecimal._
+  import Multiplication._
+
+  /** The <code>String</code> representation is cached. */
+  private var ToStringImage: String = null
```

Comment:
As var's, they should indeed start lower case. However, then there is the issue with methods having the same name. Are you OK with these private vars being prefixed with an underscore as per [the scala style guide](http://docs.scala-lang.org/style/naming-conventions.html)? e.g `_precision`